src/lib.rs

```diff
@@ -1,8 +1,18 @@
 #![no_std]
 #![cfg_attr(feature = "nightly", feature(core_intrinsics))]
 
-#[macro_use]
-extern crate log;
-
 pub mod hardware;
 pub mod net;
+
+/// Macro to reduce rightward drift when calling the same closure-based API
+/// on multiple structs simultaneously, e.g. when accessing DMA buffers.
+/// This could be improved a bit using the tuple-based style from `mutex-trait`.
+#[macro_export]
+macro_rules! flatten_closures {
+    ($fn:ident, $e:ident, $fun:block) => {
+        $e.$fn(|$e| $fun ).unwrap()
+    };
+    ($fn:ident, $e:ident, $($es:ident),+, $fun:block) => {
+        $e.$fn(|$e| flatten_closures!($fn, $($es),*, $fun)).unwrap()
```

Comment:
Have you done a forced failure test to ensure that the flag check unwrap actually fails if the DSP process is taking too long? It would be good to verify that this is still caught.

It also occurs to me that we can now ignore the ADC0/ADC1 FIFO ISRs because we're offloading that logic to this unwrap here.

Reply:
Follow-up: It looks like you indicate this in your original post. Closing as resolved.